x-pack/plugins/lens/common/expressions/time_scale/time_scale.ts

```diff
@@ -0,0 +1,151 @@
+  ExpressionFunctionDefinition,
+  Datatable,
+} from '../../../../../../src/plugins/expressions/common';
+import {
+  getDateHistogramMetaDataByDatatableColumn,
+  parseInterval,
+  calculateBounds,
+} from '../../../../../../src/plugins/data/common';
+import { buildResultColumns } from '../../../../../../src/plugins/expressions/common';
+import type { TimeScaleUnit } from './types';
+
+export interface TimeScaleArgs {
+  dateColumnId: string;
+  inputColumnId: string;
+  outputColumnId: string;
+  targetUnit: TimeScaleUnit;
+  outputColumnName?: string;
+}
+
+const unitInMs: Record<TimeScaleUnit, number> = {
+  s: 1000,
+  m: 1000 * 60,
+  h: 1000 * 60 * 60,
+  d: 1000 * 60 * 60 * 24,
+};
+
+export const timeScale: ExpressionFunctionDefinition<
+  'lens_time_scale',
+  Datatable,
+  TimeScaleArgs,
+  Promise<Datatable>
+> = {
+  name: 'lens_time_scale',
+  type: 'datatable',
+  help: '',
+  args: {
+    dateColumnId: {
+      types: ['string'],
+      help: '',
+      required: true,
+    },
+    inputColumnId: {
+      types: ['string'],
+      help: '',
+      required: true,
+    },
+    outputColumnId: {
+      types: ['string'],
+      help: '',
+      required: true,
+    },
+    outputColumnName: {
+      types: ['string'],
+      help: '',
+    },
+    targetUnit: {
+      types: ['string'],
+      options: ['s', 'm', 'h', 'd'],
+      help: '',
+      required: true,
+    },
+  },
+  inputTypes: ['datatable'],
+  async fn(
+    input,
+    { dateColumnId, inputColumnId, outputColumnId, outputColumnName, targetUnit }: TimeScaleArgs
+  ) {
+    const dateColumnDefinition = input.columns.find((column) => column.id === dateColumnId);
+
+    if (!dateColumnDefinition) {
+      throw new Error(
+        i18n.translate('xpack.lens.functions.timeScale.dateColumnMissingMessage', {
+          defaultMessage: 'Specified dateColumnId {columnId} does not exist.',
+          values: {
+            columnId: dateColumnId,
+          },
+        })
+      );
+    }
+
+    const resultColumns = buildResultColumns(
+      input,
+      outputColumnId,
+      inputColumnId,
+      outputColumnName,
+      { allowColumnOverwrite: true }
+    );
+
+    if (!resultColumns) {
+      return input;
+    }
+
+    const targetUnitInMs = unitInMs[targetUnit];
+    const timeInfo = getDateHistogramMetaDataByDatatableColumn(dateColumnDefinition);
+    const intervalDuration = timeInfo?.interval && parseInterval(timeInfo.interval);
+
+    if (!timeInfo || !intervalDuration) {
+      throw new Error(
+        i18n.translate('xpack.lens.functions.timeScale.timeInfoMissingMessage', {
+          defaultMessage: 'Could not fetch date histogram information',
+        })
+      );
+    }
+    // the datemath plugin always parses dates by using the current default moment time zone.
+    // to use the configured time zone, we are switching just for the bounds calculation.
+    const defaultTimezone = moment().zoneName();
+    moment.tz.setDefault(timeInfo.timeZone);
```

Comment:
What I'm seeing is that `timeInfo.timeZone` is undefined most of the time, because the time zone is not provided on esaggs by default. The time zone information might need to be sourced from an alternative place. For example, if the alerting task starts it, it might need to store a timezone parameter that gets passed in as part of the expression context. Have you considered that here?

Reply:
I've created a new argument for `time_scale` to pass a `timeZone` to be used on the server side: you think it would be better to extend the context instead?

Reply:
I'm not sure how this will be handled, I think it will require discussion with the app services & alerting teams.

Reply:
Left a comment on the main PR thread: this topic will be discussed in a follow up PR completely dedicated to the server side registration.